From 6f6820f690629e675dcb05bfc8a1f287f0a98641 Mon Sep 17 00:00:00 2001
From: Joe Rafaniello <jrafanie@gmail.com>
Date: Fri, 3 Dec 2010 14:17:51 -0500
Subject: [PATCH] Don't silently discard exceptions when getting/setting session via memcache.

This was fixed in commit d0df7f2b12b201edbef0d59360224e68f9a0a447 old trac #9823 and then broken with commit ed708307137c811d14e5fd2cb4ea550add381a82 when switching to rack based session stores.
---
 .../action_controller/session/mem_cache_store.rb   |   10 ++--------
 .../controller/session/mem_cache_store_test.rb     |   14 ++++++++++++++
 2 files changed, 16 insertions(+), 8 deletions(-)

diff --git a/actionpack/lib/action_controller/session/mem_cache_store.rb b/actionpack/lib/action_controller/session/mem_cache_store.rb
index 402681c..3ab8644 100644
--- a/actionpack/lib/action_controller/session/mem_cache_store.rb
+++ b/actionpack/lib/action_controller/session/mem_cache_store.rb
@@ -27,11 +27,7 @@ begin
         private
           def get_session(env, sid)
             sid ||= generate_sid
-            begin
-              session = @pool.get(sid) || {}
-            rescue MemCache::MemCacheError, Errno::ECONNREFUSED
-              session = {}
-            end
+            session = @pool.get(sid) || {}
             [sid, session]
           end
 
@@ -39,9 +35,7 @@ begin
             options = env['rack.session.options']
             expiry  = options[:expire_after] || 0
             @pool.set(sid, session_data, expiry)
-            return true
-          rescue MemCache::MemCacheError, Errno::ECONNREFUSED
-            return false
+            true
           end
           
           def destroy(env)
diff --git a/actionpack/test/controller/session/mem_cache_store_test.rb b/actionpack/test/controller/session/mem_cache_store_test.rb
index 2714ee5..c4243e7 100644
--- a/actionpack/test/controller/session/mem_cache_store_test.rb
+++ b/actionpack/test/controller/session/mem_cache_store_test.rb
@@ -12,6 +12,12 @@ class MemCacheStoreTest < ActionController::IntegrationTest
       head :ok
     end
 
+    ONE_MB = 1024 * 1024
+    def set_one_mb_session_value
+      session[:foo] = "a" * ONE_MB
+      head :ok
+    end
+
     def set_serialized_session_value
       session[:foo] = SessionAutoloadTest::Foo.new
       head :ok
@@ -50,6 +56,14 @@ class MemCacheStoreTest < ActionController::IntegrationTest
       end
     end
 
+    def test_setting_one_mb_session_value_raises_exception
+      with_test_route_set do
+        assert_raise(MemCache::MemCacheError) do
+          get '/set_one_mb_session_value'
+        end
+      end
+    end
+
     def test_getting_nil_session_value
       with_test_route_set do
         get '/get_session_value'
-- 
1.7.3.2

